File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.cpp
class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& cls, int extra) {
        priority_queue< pair< double, vector<int>>> maxHeap;

        for(auto &cl: cls)
        {
            int pass = cl[0];
            int total = cl[1];

            double gain = (double)(pass+1)/(total+1) - (double)(pass)/total;
            maxHeap.push({gain, cl});
        }

        while(extra--)
        {
            auto top = maxHeap.top();
            maxHeap.pop();

            vector<int> cl = top.second;

            int pass = cl[0];
            int total = cl[1];

            pass++;
            total++;

            cl = {pass, total};

            // new gain
            double gain = (double)(pass+1)/(total+1) - (double)(pass)/total;
            maxHeap.push({gain, cl});
        }

        double totalSum = 0.0;
        while(!maxHeap.empty())
        {
            auto top = maxHeap.top();
            maxHeap.pop();

            vector<int> cl = top.second;

            int pass = cl[0];
            int total = cl[1];

            totalSum += (double)pass/total;
        }

        return totalSum / cls.size();
    }
};

File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.java
import java.util.PriorityQueue;

class Solution {
    public double maxAverageRatio(int[][] cls, int extra) {
        PriorityQueue<double[]> maxHeap = new PriorityQueue<>((a, b) -> Double.compare(b[0], a[0]));

        for (int[] cl : cls) {
            int pass = cl[0], total = cl[1];
            double gain = (double)(pass + 1) / (total + 1) - (double)pass / total;
            maxHeap.offer(new double[]{gain, pass, total});
        }

        while (extra-- > 0) {
            double[] top = maxHeap.poll();
            double gain = top[0];
            int pass = (int)top[1];
            int total = (int)top[2];

            pass++;
            total++;

            double newGain = (double)(pass + 1) / (total + 1) - (double)pass / total;
            maxHeap.offer(new double[]{newGain, pass, total});
        }

        double totalSum = 0.0;
        while (!maxHeap.isEmpty()) {
            double[] top = maxHeap.poll();
            int pass = (int)top[1];
            int total = (int)top[2];
            totalSum += (double)pass / total;
        }

        return totalSum / cls.length;
    }
}
